pkg/backends/tcp_test.go

```diff
@@ -56,112 +167,658 @@ func TestNewTCPListener(t *testing.T) {
 	}
 }
 
-func TestTCPListenerStart(t *testing.T) {
-	type fields struct {
-		address string
-		TLS     *tls.Config
-		li      net.Listener
-		innerLi *net.TCPListener
-		logger  *logger.ReceptorLogger
+func TestTCPListener_GetAddr(t *testing.T) {
+	tests := []struct {
+		name         string
+		address      string
+		expectedHost string
+	}{
+		{
+			name:         "IPv4 address with port 0",
+			address:      "127.0.0.1:0",
+			expectedHost: "127.0.0.1",
+		},
+		{
+			name:         "IPv6 address with port 0",
+			address:      "[::1]:0",
+			expectedHost: "::1",
+		},
 	}
-	type args struct {
-		ctx context.Context
-		wg  *sync.WaitGroup
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			b, err := NewTCPListener(tt.address, nil, logger.NewReceptorLogger("TCPtest"))
+			if err != nil {
+				t.Errorf("NewTCPListener() error = %v", err)
+
+				return
+			}
+
+			// Address is bound after listener start
+			ctx, cancel := context.WithCancel(context.Background())
+			wg := &sync.WaitGroup{}
+			defer func() {
+				cancel()
+				wg.Wait()
+			}()
+			_, err = b.Start(ctx, wg)
+			if err != nil {
+				t.Errorf("TCPListener.Start() error = %v", err)
+
+				return
+			}
+
+			got := b.GetAddr()
+			if got == "" {
+				t.Errorf("TCPListener.GetAddr() returned empty string")
+
+				return
+			}
+
+			host, _, err := net.SplitHostPort(got)
+			if err != nil {
+				t.Errorf("TCPListener.GetAddr() returned invalid address format %q: %v", got, err)
+
+				return
+			}
+
+			if host != tt.expectedHost {
+				t.Errorf("TCPListener.GetAddr() host = %v, want %v", host, tt.expectedHost)
+			}
+		})
+	}
+}
+
+func TestTCPListener_GetTLS(t *testing.T) {
+	tlsConfig := &tls.Config{
+		MinVersion: tls.VersionTLS12,
 	}
+
+	tests := []struct {
+		name string
+		tls  *tls.Config
+		want *tls.Config
+	}{
+		{
+			name: "No TLS",
+			tls:  nil,
+			want: nil,
+		},
+		{
+			name: "With TLS config",
+			tls:  tlsConfig,
+			want: tlsConfig,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			b, err := NewTCPListener("127.0.0.1:0", tt.tls, nil)
+			if err != nil {
+				t.Errorf("NewTCPListener() error = %v", err)
+
+				return
+			}
+			if got := b.GetTLS(); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("TCPListener.GetTLS() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestTCPListenerStart(t *testing.T) {
 	tests := []struct {
 		name    string
-		fields  fields
-		args    args
-		want    chan netceptor.BackendSession
+		address string
+		tls     *tls.Config
 		wantErr bool
 	}{
 		{
-			name: "Positive",
-			fields: fields{
-				address: "127.0.0.1:9998",
-				TLS:     nil,
-				li:      nil,
-				innerLi: nil,
-				logger:  logger.NewReceptorLogger("TCPtest"),
-			},
-			args: args{
-				ctx: context.Background(),
-				wg:  &sync.WaitGroup{},
-			},
+			name:    "Successful start",
+			address: "127.0.0.1:9998",
```

Comment:
```suggestion
			address: "127.0.0.1:0",
```
Same thing as before.